### PR TITLE
Push message and background data support.

### PR DESCRIPTION
**Pull request by zyro**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,13 +77,7 @@ GameUp.login(this, new GameUpLoginListener() {
...
 You can check the service is live and the `GameUpSession` and API key are valid:
 ```java
-final boolean pingResult = gameUpSession.ping();
-if(pingResult) {
-    // Service is reachable and session/key are valid.
-}
-else {
-    // Something went wrong.
-}
+gameUpSession.ping();
```
Comment (novabyte) [inline README.md]: What does `ping()` do if it fails? Should we mention how to capture the failure?

File: README.md [hunk 2]
```diff
@@ -43,12 +43,18 @@ The login flow is handled automatically in a `WebView`, with a callback invoked
...
+GameUp.init(context, "your-api-key");
 GameUp.login(this, new GameUpLoginListener() {
     @Override
     public void onLoginComplete(final GameUpSession gameUpSession) {
         // Called with the new session when login completes successfully.
-        // You may want to:
+
+        // You should always update push messaging registration.
+        // Push subscription settings might come from user options.
+        gameUpSession.push(context, true, Arrays.asList("some", "segments"));
```
Comment (novabyte) [inline README.md]: Should we use more descriptive segment names here; maybe "REWARDS_FROM_ARENA", "STAMINA_FULL", etc?

File: sdk/src/main/java/io/gameup/android/json/PushRegistration.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package io.gameup.android.json;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/json/PushRegistration.java]: Missing License header.

File: sdk/src/main/java/io/gameup/android/json/PushRegistration.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.List;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+/**
+ * Created by andrei on 20/04/15.
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/json/PushRegistration.java]: Needs proper javadoc comments.

File: sdk/src/main/java/io/gameup/android/push/GcmBroadcastReceiver.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package io.gameup.android.push;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/GcmBroadcastReceiver.java]: Missing License header.

File: sdk/src/main/java/io/gameup/android/push/GcmIntentService.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package io.gameup.android.push;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/GcmIntentService.java]: Needs License header.

File: sdk/src/main/java/io/gameup/android/push/GcmIntentService.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    protected void onHandleIntent(final Intent intent) {
+        if (isGcmMessage(intent)) {
+            NotificationBundleProcessor.process(this, intent.getExtras());
+        }
+
+        // Release the wake lock provided by the WakefulBroadcastReceiver.
+        GcmBroadcastReceiver.completeWakefulIntent(intent);
+    }
+
+}
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/GcmIntentService.java]: Missing newline.

File: sdk/src/main/java/io/gameup/android/push/GenerateNotification.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
+package io.gameup.android.push;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/GenerateNotification.java]: Missing License header.

File: sdk/src/main/java/io/gameup/android/push/GenerateNotification.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            notificationDefaults |= Notification.DEFAULT_SOUND;
+        }
+
+        mBuilder.setDefaults(notificationDefaults);
+        mBuilder.setContentIntent(contentIntent);
+
+        mNotificationManager.notify(notificationId, mBuilder.build());
+    }
+
+}
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/GenerateNotification.java]: Missing newline.

File: sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+                intent.putExtra("data", bundle);
+
+                context.sendBroadcast(intent);
+            }
+
+            // If there is no alert data, skip the notification.
+            if (!bundle.containsKey("alert") ||
+                    bundle.getString("alert") == null ||
+                    bundle.getString("alert").equals(""))
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java]: Missing braces around conditional blocks.

File: sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
+package io.gameup.android.push;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java]: Missing License header.

File: sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Format short keys into more readable ones.
+    private static void prepareBundle(final Bundle gcmBundle) {
+        if (gcmBundle.containsKey("o")) {
+            try {
+                final JSONObject customJSON = new JSONObject(
+                        gcmBundle.getString("custom"));
+                JSONObject additionalDataJSON;
+
+                if (customJSON.has("a"))
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java]: Missing braces around conditional blocks.

File: sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+                    customJSON.put("a", additionalDataJSON);
+
+                gcmBundle.putString("custom", customJSON.toString());
+            } catch (final JSONException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+}
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/NotificationBundleProcessor.java]: Missing newline.

File: sdk/src/main/java/io/gameup/android/push/NotificationOpenedActivity.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
+package io.gameup.android.push;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/NotificationOpenedActivity.java]: Missing License header.

File: sdk/src/main/java/io/gameup/android/push/NotificationOpenedActivity.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+        super.onNewIntent(intent);
+        processNotification();
+    }
+
+    private void processNotification() {
+        Push.handleNotificationOpened(this, getIntent().getBundleExtra("data"));
+        finish();
+    }
+
+}
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/NotificationOpenedActivity.java]: Missing newline.

File: sdk/src/main/java/io/gameup/android/push/Push.java [hunk 16]
```diff
@@ -0,0 +1,231 @@
+package io.gameup.android.push;
```
Comment (novabyte) [inline sdk/src/main/java/io/gameup/android/push/Push.java]: Missing License header.

File: test-application/src/main/java/io/gameup/android/test/PushBackgroundReceiver.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+                            .getJSONObject("a");
+
+            Toast.makeText(context,
+                    "BACKGROUND, data=" + data.toString(),
+                    Toast.LENGTH_LONG).show();
+        } catch (Throwable t) {
+            t.printStackTrace();
+        }
+    }
+}
```
Comment (novabyte) [inline test-application/src/main/java/io/gameup/android/test/PushBackgroundReceiver.java]: Missing newline.

File: test-application/src/main/java/io/gameup/android/test/PushBackgroundReceiver.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
+package io.gameup.android.test;
```
Comment (novabyte) [inline test-application/src/main/java/io/gameup/android/test/PushBackgroundReceiver.java]: Missing License header.

**Comment by novabyte** [comment]
@zyro Mostly cosmetic changes but I get a dependency error when I try to build the code:

```shell
A problem occurred configuring project ':sdk'.
> Failed to notify project evaluation listener.
   > Could not resolve all dependencies for configuration ':sdk:_debugCompile'.
      > Could not find com.google.android.gms:play-services:7.0.0.
        Searched in the following locations:
            https://repo1.maven.org/maven2/com/google/android/gms/play-services/7.0.0/play-services-7.0.0.pom
            https://repo1.maven.org/maven2/com/google/android/gms/play-services/7.0.0/play-services-7.0.0.jar
            file:/usr/local/opt/android-sdk/extras/android/m2repository/com/google/android/gms/play-services/7.0.0/play-services-7.0.0.pom
            file:/usr/local/opt/android-sdk/extras/android/m2repository/com/google/android/gms/play-services/7.0.0/play-services-7.0.0.jar
        Required by:
            io.gameup.android:sdk:0.5.0-SNAPSHOT
```

**Comment by zyro** [comment]
Make sure you have the latest build tools and SDK installed, especially:
screen shot 2015-04-26 at 12 20 26

**Comment by novabyte** [comment]
@zyro Awesome :+1:

**Pull request closed without merging**